File: Python/0001-0050/22. Generate Parentheses.py
# https://leetcode.com/problems/generate-parentheses/description/

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def generateRecursively(opening, closing, current):
            if len(current) == n*2:
                result.append(current)
                return

            # Recursive calls
            if opening < n:
                generateRecursively(opening+1, closing, current+"(")
            
            if closing < opening:
                generateRecursively(opening, closing+1, current+")")
            
        result = []
        generateRecursively(0,0,"")
        return result

File: Java/0001-0050/22. Generate Parentheses.java
// https://leetcode.com/problems/generate-parentheses/description/

class Solution {
    public void generateRecursively(List<String> result,int n, int opening, int closing, String current){
        if(current.length() == n*2){
            result.add(current);
            return;
        }

        // Recursive calls
        if(opening < n){
            generateRecursively(result,n, opening+1, closing, current+"(");
        }
        if(closing < opening){
            generateRecursively(result,n, opening, closing+1, current+")");
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        generateRecursively(result, n, 0, 0, "");
        return result;
    }
}
